### ISSUE TITLE
[논문설명] Using Adversarial Autoencoders for Multi-Modal Automatic Playlist Continuation(1) - VAE

### ISSUE THREAD

**Issue by jayden5744** [jayden5744/AE#1]
# 1. Variational AutoEncoder(VAE)의 목적
- AutoEncoder의 목적과 같이 주어진 데이터를 잘 설명하는 latent vector 즉, z를 찾는 것이다.
- 그러나 차이점은 VAE는 고정된 z가 아닌 z의 분포를 찾는 것이다. 
- 쉽게 말하면 AE의 z는 원본 x를 제일 잘 표현하는 값을 찾는 방식이고, VAE는 원본 x를 제일 잘 표현하는 분포를 찾는 방식이다. 

### 값이 아닌 분포를 해서 얻을 수 있는 이점은 뭘까?
- 아무런 가정 없이 random variable만 입력으로 넣어주고 데이터를 복원해보라고하면 너무 어렵고 잘 될리가 만무하니까 잘 복원해주는 random variable을 만들 수는 없을까라는 고민에서 비롯한 방법
- 그렇다고 해서 AE보다 VAE가 더 좋다고는 할 수 없고, 다만 학습은 더 안정적으로 된다고 함.

![image](https://user-images.githubusercontent.com/47970983/84150857-6f2c8f00-aa9d-11ea-85ea-fca39c130844.png)

- 먼저 분포를 정의하기 위해서는 모수를 정의해야 하는데, 여기서는 분포를 정규분포로 가정하기 때문에 평균과 표준 편차를 모수로 가진다.
- Encoder는 학습용 데이터(x)를 입력으로 받고 잠재 변수(z)의 확률분포에 대한 파라미터(모수)를 출력한다
- Decoder 는 잠재변수에 대한 확률 분포 p(z)에서 샘플링한 벡터를 입력받아 이를 이용해 원본 데이터를 복원한다.

# 2. Encoder 
- Encoder의 역할은 데이터가 주어졌을 때 Decoder가 원래의 데이터로 잘 복원할 수 있는 z를 샘플링할 수 있는 이상적인 p(z|x)를 찾는 것
- 그러나 이상적인 p(z|x)가 무엇인지 아무도 모른다
- 이를 알기 위한 방법으로 Variational inference를 사용한다

## 2-1 Variational inference
- 우리가 이상적인 확률분포를 모를 때, 이를 추정하기 위해 다루기 쉬운 분포를 가정하고 이 확률분포의 모수를 바꿔가며, 이상적인 확률분포에 근사하게 만들어 그 확률분포를 대신 사용하는 것
- 이 다루기 쉬운 분포를 qϕ 라고 한다면, Encoder는 ϕ 라는 파라미터(모수)들을 바꾸어가며, qϕ(z|x) 확률 분포를 이상적인 확률 분포 p(z|x) 에 근사시키는 역할을 수행한다
![image](https://user-images.githubusercontent.com/47970983/84151428-54a6e580-aa9e-11ea-931b-1eceac34d693.png)
- Encoder의 output은 qϕ(z|x) 확률 분포의 모수이다. 위에서 우리는 **정규분포로 가정**했기 때문에, **확률분포의 모수는 평균과 분산**이다.
- 평균과 분산을 가지고 그냥 샘플링 한다면 back propagation이 불가능하다. 그래서 사용하는 것이 **Reparameterization trick**이다.

### 왜 back propagation이 불가능한가?
- 밑에 그림을 보면 Deterministic node는 Backpropagation이 가능하지만 Random Node는 그럴 수가 없다. 왜냐하면 어떤 수식으로 인하여 나온 결과를 역으로 추적하는 것이 Back propagation인데 Random으로 나온 결과는 수식에 따라 타고 갈 수 없기 떄문에(뇌피셜)
![image](https://user-images.githubusercontent.com/47970983/84151999-11994200-aa9f-11ea-9fde-2d6564f571f5.png)

## 2-2 Reparameterization trick
- 가우시안 정규분포의 샘플을 추출하고 싶을 때, 아래의 식과 같이 샘플링 하는 것을 말함
- 확률적 특성을 그대로 보존하면서, Back propagation이 가능함

![image](https://user-images.githubusercontent.com/47970983/84152457-a13ef080-aa9f-11ea-87f7-5bb17d71d9a5.png)

![image](https://user-images.githubusercontent.com/47970983/84152506-b9af0b00-aa9f-11ea-856d-76ed3b8cdd7d.png)

``` python

import torch

def __init__(self):
    self.fc1_1 = nn.Linear(784, hidden_size)
    self.fc1_2 = nn.Linear(784, hidden_size)
    self.relu = nn.ReLU()
                        
def encode(self,x):
    x = x.view(batch_size,-1)
    mu = self.relu(self.fc1_1(x))     # mean vector
    log_var = self.relu(self.fc1_2(x))  # variation
    return mu,log_var
    
def reparametrize(self, mu, logvar):  # Reparametrization Trick
    std = logvar.mul(0.5).exp_()      # σ(x)
    eps = torch.FloatTensor(std.size()).normal_()  # ϵ∼N(0,1)
    return eps.mul(std).add_(mu)     # z=μ(x)+σ(x)×ϵ
```

# 3. Decoder 
- Decoder는 추출한 샘플(z)를 입력으로 받아 원본을 재구축하는 역할을 한다. 기존의 AutoEncoder의 Decoder와 동일

# 4. Loss Function - Evidence LowerBOund(ELBO)
- 우리가 알고 싶은 것은 p(x)이다.

## 4-1 Jensen's Inquality
![image](https://user-images.githubusercontent.com/47970983/84154045-8b322f80-aaa1-11ea-9363-f4f04ff0c041.png)

![image](https://user-images.githubusercontent.com/47970983/84154164-b87edd80-aaa1-11ea-8df6-574425581169.png)

``` python 
# the Binary Cross Entropy between the target and the output
reconstruction_function = nn.BCELoss(size_average=False)

def loss_function(recon_x, x, mu, logvar):
    BCE = reconstruction_function(recon_x, x)
    KLD_element = mu.pow(2).add_(logvar.exp()).mul_(-1).add_(1).add_(logvar)
    KLD = torch.sum(KLD_element).mul_(-0.5)
    return BCE + KLD
```
- BCE로 input x와 output x를 비교해서 loss를 구하고
- KLD로 정규분포랑 원래 z분포의 loss를 구한다.

# 참고자료
- 논문 : https://zenodo.org/record/1455214/files/Vagliano-et-al-Using-Adversarial-Autoencoders-for-Multi-Modal-Automatic-Playlist-Continuation.pdf?download=1
- 코드 : https://github.com/lgalke/mpd-aae-recommender
- 블로그
    - https://ratsgo.github.io/generative%20model/2018/01/27/VAE/
    - https://datascienceschool.net/view-notebook/c5248de280a64ae2a96c1d4e690fdf79/
    - https://www.youtube.com/watch?v=-iamEYdDyvE
